crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -10559,4 +10570,27 @@ mod tests {
 
         Ok(())
     }
+
+    // Regression test for https://github.com/astral-sh/ty/issues/649
+    #[test]
+    fn bitwise_and_with_nominal_instance() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "/src/a.py",
+            r#"
+            from typing_extensions import assert_type
+            from random import random
+
+            def f() -> bool:
+                return random() > 0.5
+
+            assert_type(True & f(), bool)
+            "#,
+        )?;
+
+        assert_file_diagnostics(&db, "src/a.py", &[]);
+
+        Ok(())
+    }
```

Comment:
We write all such tests as mdtests (test snippets embedded as Markdown) rather than as regular Rust tests. See https://github.com/astral-sh/ruff/blob/main/crates/ty/CONTRIBUTING.md#writing-tests for details. This test should be easy to transform to an mdtest.

Reply:
Moved 👍🏻